### PR TITLE
Handle research trigger 'craft-item'

### PR DESCRIPTION
**Pull request by veger**
While checking for the Space Age changes of #320, I noticed 'new' errors when loading a project. Which was triggered by the (lack of) support of the new [research triggers](https://lua-api.factorio.com/latest/concepts/ResearchTrigger.html).

Adding the `craft-item` type was fairly easy (I was fiddling a bit and it worked):
_\<removed outdated screenshot, see comments for newer one\>_

So I cleaned the code a little and made this PR.

Note that supporting `mine-entity` was also easy, but for some reason it makes most of the items inaccessible... So I did not include it in this PR. We need to take a better look.
The other types, I did not try as their implementation was not directly apparent to me (and `craft-fluid` is not in use, so I could not try/test)

### CODE REVIEW

File: Yafc.Parser/Data/FactorioDataDeserializer_RecipeAndTechnology.cs [hunk 1]
```diff
@@ -229,6 +229,29 @@ private Ingredient[] LoadResearchIngredientList(LuaTable table, string typeDotNa
...
+            return;
+        }
+
+        switch (type) {
+            case "craft-item":
+                if (!researchTriggerTable.Get("item", out string? craftItemName)) {
+                    errorCollector.Error($"Research trigger craft-item of {technology.typeDotName} does not have a item field", ErrorSeverity.MinorDataLoss);
+                    break;
+                }
+                researchTriggerTable.Get("count", out float craftCount, 1);
```
Comment (SWeini) [inline Yafc.Parser/Data/FactorioDataDeserializer_RecipeAndTechnology.cs:244]: minor code style - I don't know whether we even have some conventions for it, but I have seen too many out parameters lately

`float craftCount = researchTriggerTable.Get("count", 1f);` wouldn't need the out parameter, which is always a bit awkward and doesn't discard the bool result

**Comment by SWeini** [comment]
Thanks for doing the UI work.

I don't have anything against this change, but I want to explore the topic of trigger techs a bit further:

For me the big challenge was to include the triggers into the tech tree / milestone analysis. Unfortunately it doesn't build up one big graph and then "simply" walks along edges and marks nodes as "available". It has some really weird logic. So I went the easy way, basically enabled all techs (by binding `SpecialNames.TechnologyTrigger` to the character) and hoped that the rest of the tech tree simply contains enough information for everything else to work. For vanilla that was OK, but the mods we need a better long-term solution.

e.g. the [MineEntityTechnologyTrigger](https://lua-api.factorio.com/latest/types/MineEntityTechnologyTrigger.html) trigger needs a lot of requirements:
- if entity is non-resources entity (I guess this works, but I haven't checked): character needs access to a surface/planet where that entity can auto-spawn, or player is able to build one itself
- if entity is resource:
  - character needs access to a surface/planet where that resource can auto-spawn
  - character can mine its [category](https://lua-api.factorio.com/latest/prototypes/ResourceEntityPrototype.html#category) or has access to a mining drill that can, and access to fuel for that drill, and the drill is allowed on that same surface/planet where the resource is
  - we need to respect [MiningWithFluidModifier](https://lua-api.factorio.com/latest/types/MiningWithFluidModifier.html), which we don't support at all yet

To sum it up, I'm not surprised that the `mine-entity` trigger didn't work for you out of the box - and we need to think about how much brainpower we put into the trigger techs and how we prioritize it.

**Comment by SWeini** [comment]
is it possible to also remove the ingredients that show up the "normal way"

image

File: Yafc.Parser/Data/FactorioDataDeserializer_RecipeAndTechnology.cs (hunk 1 above)
Comment (veger) [reply Yafc.Parser/Data/FactorioDataDeserializer_RecipeAndTechnology.cs:244]: I just copied from another place as there are so/too many variants of the `LuaTable.Get()` function.

**Comment by veger** [comment]
I removed the 'recipe part' of the card when showing a research trigger technology, as all of the recipe details did not make much sense for the them:
image

**Comment by veger** [comment]
@SWeini we could add another field to the `Technology` class instead of (ab)using the `RecipeOrTechnology.ingredients` field... This might prevent unforeseen/future issues...? :thinking:

**Comment by shpaass** [comment]
Rebased on fresh master

**Review by shpaass — APPROVED** [review]

File: Yafc.Parser/Data/FactorioDataDeserializer_RecipeAndTechnology.cs (hunk 1 above)
Comment (shpaass) [reply Yafc.Parser/Data/FactorioDataDeserializer_RecipeAndTechnology.cs]: Resolving this conversation to merge the PR. Please reopen if necessary.
